src/realm/deppart/byfield_gpu_impl.hpp

```diff
@@ -0,0 +1,268 @@
+#pragma once
+#include "realm/deppart/byfield.h"
+#include "realm/deppart/byfield_gpu_kernels.hpp"
+#include "realm/deppart/partitions_gpu_impl.hpp"
+#include <cub/cub.cuh>
+#include "realm/nvtx.h"
+
+namespace Realm {
+
+//Used to compute prefix sum by volume for an array of rects
+template <int N, typename T>
+struct RectFieldVolumeOp {
+  __device__ __forceinline__
+  size_t operator()(const Rect<N,T>& r) const {
+    return r.volume();
+  }
+};
+
+/*
+ *  Input (stored in MicroOp): Array of field instances, a parent index space, and a list of colors
+ *  Output: A list of (potentially overlapping) points in original instances ∩ parent index space marked with their color,
+ *  which it then sends off to complete_pipeline.
+ *  Approach: Intersect all instance rectangles with parent rectangles in parallel. For surviving rectangles, use
+ *  prefix sum + binary search to iterate over these in parallel and mark each point with its color.
+ */
+template <int N, typename T, typename FT>
+void GPUByFieldMicroOp<N,T,FT>::execute()
+{
+
+    NVTX_DEPPART(byfield_gpu);
+
+    cudaStream_t stream = Cuda::get_task_cuda_stream();
+
+    Memory my_mem = field_data[0].inst.get_location();
+
+
+    // We need inst_offsets to preserve which instance each rectangle came from
+    // so that we know where to read from after intersecting.
+    std::vector<size_t> inst_offsets(field_data.size() + 1);
+
+    //Determine size of allocation for combined inst_rects and offset entries
+    size_t inst_size = 0;
+
+
+    for (size_t i = 0; i < field_data.size(); ++i) {
+      inst_offsets[i] = inst_size;
+      if (field_data[i].index_space.dense()) {
+        inst_size += 1;
+      } else {
+        inst_size += field_data[i].index_space.sparsity.impl()->get_entries().size();
+      }
+    }
+    inst_offsets[field_data.size()] = inst_size;
+
+
+    RegionInstance inst_entries_instance = this->realm_malloc(inst_size * sizeof(SparsityMapEntry<N,T>), my_mem);
```

Comment:
If I remember the presentation correctly, could this first group of work (intersecting source space rectangles etc) be extracted into a separate helper that the other ops will use too?